### ISSUE TITLE
Box::default recommendation is overly broad (rustc can't infer type)

### ISSUE THREAD

**Issue by dacut** [rust-lang/rust-clippy#10381]
### Summary

Clippy's box_default lint rule is overly optimistic. It recommends replacing instances of `Box::new(X::default())` with `Box::default()` in cases where the compiler can't infer the type. For example, if `X` implements a trait `T` and the return type is `Box<dyn T>` instead of `Box<X>`, there's no way for rustc to figure out what should be done here.

### Lint Name

box_default

### Reproducer

I tried this code:

```rust
#[derive(Default)]
pub struct Foo {}
pub trait Bar {}
impl Bar for Foo {}

fn maybe_get_bar(i: u32) -> Option<Box<dyn Bar>> {
    if i % 2 == 0 {
        Some(Box::new(Foo::default()))
    } else {
        None
    }
}
```

I saw this happen:

```
    Checking playground v0.0.1 (/playground)
warning: `Box::new(_)` of default value
 --> src/main.rs:8:14
  |
8 |         Some(Box::new(Foo::default()))
  |              ^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `Box::default()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#box_default
  = note: `#[warn(clippy::box_default)]` on by default

warning: `playground` (bin "playground") generated 1 warning (run `cargo clippy --fix --bin "playground"` to apply 1 suggestion)
    Finished dev [unoptimized + debuginfo] target(s) in 0.58s
```

I expected to see this happen:
No Clippy warnings.

Replacing `Box::new(Foo::default())` with `Box::default()`, of course, produces this error:
```
Compiling playground v0.0.1 (/playground)
error[[E0277]](https://doc.rust-lang.org/nightly/error_codes/E0277.html): the trait bound `dyn Bar: Default` is not satisfied
 --> src/main.rs:8:14
  |
8 |         Some(Box::default())
  |              ^^^^^^^^^^^^ the trait `Default` is not implemented for `dyn Bar`
  |
  = help: the following other types implement trait `Default`:
            Box<CStr>
            Box<OsStr>
            Box<T>
            Box<[T]>
            Box<str>
  = note: required for `Box<dyn Bar>` to implement `Default`

For more information about this error, try `rustc --explain E0277`.
error: could not compile `playground` due to previous error
```

### Version

```text
Locally:
rustc 1.69.0-nightly (75a0be98f 2023-02-05)
binary: rustc
commit-hash: 75a0be98f25a4b9de5afa0e15eb016e7f9627032
commit-date: 2023-02-05
host: aarch64-apple-darwin
release: 1.69.0-nightly
LLVM version: 15.0.7

And on Playground:
Build using the Nightly version: 1.69.0-nightly
(2023-02-19 7aa413d59206fd511137)
```
